src/storage/src/source/postgres/snapshot.rs

```diff
@@ -0,0 +1,376 @@
+
+            let client = connection_config.connect_replication().await?;
+            if is_snapshot_leader {
+                // The main slot must be created *before* we start snapshotting so that we can be
+                // certain that the temporarly slot created for the snapshot start at an LSN that
+                // is greater than or equal to that of the main slot.
+                super::ensure_replication_slot(&client, &connection).await?;
+
+                let snapshot_info = export_snapshot(&client).await?;
+                trace!(%id, "timely-{worker_id} exporting snapshot info {snapshot_info:?}");
+                let cap = snapshot_cap.as_ref().unwrap();
+                snapshot_handle.give(cap, snapshot_info).await;
+            }
+
+            let (snapshot, snapshot_lsn) = match snapshot_input.next_mut().await {
+                Some(AsyncEvent::Data(_, data)) => data.pop().expect("snapshot sent above"),
+                _ => unreachable!(),
+            };
+            if !is_snapshot_leader {
+                trace!(%id, "timely-{worker_id} using snapshot id {snapshot:?}");
+                use_snapshot(&client, &snapshot).await?;
+            }
+
+            // We have established a snapshot LSN so we can broadcast the rewind requests
+            for &oid in reader_table_info.keys() {
+                trace!(%id, "timely-{worker_id} producing rewind request for {oid}");
+                let req = RewindRequest { oid, snapshot_lsn };
+                rewinds_handle.give(rewind_cap.as_ref().unwrap(), req).await;
+            }
+            *rewind_cap = None;
+
+            let current_info = mz_postgres_util::publication_info(
+                &connection_config,
+                &connection.publication,
+                None,
+            )
+            .await?;
+            let current_info: BTreeMap<_, _> = current_info.into_iter().map(|t| (t.oid, t)).collect();
+
+            for (&oid, (_, expected_desc, _)) in reader_table_info.iter() {
+                // Verify the schema of this table
+                let desc = match current_info.get(&oid) {
+                    Some(current_desc) => {
+                        match expected_desc.determine_compatibility(current_desc) {
+                            Ok(()) => expected_desc,
+                            Err(err) => {
+                                let err = Err(DefiniteError::IncompatibleSchema(err.to_string()));
+                                raw_handle.give(data_cap, ((oid, err), MzOffset::from(0), 1)).await;
+                                continue;
+                            }
+                        }
+                    }
+                    None => {
+                        let err = Err(DefiniteError::TableDropped);
+                        raw_handle.give(data_cap, ((oid, err), MzOffset::minimum(), 1)).await;
+                        continue;
+                    }
+                };
```

Comment:
is it possible to do this verification in a function?

Reply:
plus 1

Reply:
Yes!